src/crab/ebpf_domain.cpp

```diff
@@ -119,243 +121,644 @@ static std::vector<linear_constraint_t> assume_bit_cst_interval(const NumAbsDoma
-                return {linear_constraint_t::FALSE()};
-            return {};
-        case Op::NE:
-            if (dst_lb_value == src_int_value && dst_ub_value == src_int_value)
-                return {linear_constraint_t::FALSE()};
-            if (dst_lb_value > src_int_value || dst_ub_value < src_int_value)
-                return {linear_constraint_t::TRUE()};
-            return {};
-        case Op::SGE:
-            if (dst_lb_value >= src_int_value)
-                return {linear_constraint_t::TRUE()};
-            if (dst_ub_value < src_int_value)
-                return {linear_constraint_t::FALSE()};
-            return {};
-        case Op::SLT:
-            if (dst_lb_value >= src_int_value)
-                return {linear_constraint_t::FALSE()};
-            if (dst_ub_value < src_int_value)
-                return {linear_constraint_t::TRUE()};
-            return {};
-        case Op::SLE:
-            if (dst_ub_value <= src_int_value)
-                return {linear_constraint_t::TRUE()};
-            if (dst_lb_value > src_int_value)
-                return {linear_constraint_t::FALSE()};
-            return {};
-        case Op::SGT:
-            if (dst_ub_value <= src_int_value)
-                return {linear_constraint_t::FALSE()};
-            if (dst_lb_value > src_int_value)
-                return {linear_constraint_t::TRUE()};
-            return {};
-        default: throw std::exception();
-        }
-    } else {
-        switch (op) {
-        case Op::EQ: return {dst_value == src_int_value};
-        case Op::NE: return {dst_value != src_int_value};
-        case Op::SGE: return {dst_value >= src_int_value};
-        case Op::SLE: return {dst_value <= src_int_value};
-        case Op::SGT: return {dst_value > src_int_value};
-        case Op::SLT: return {dst_value < src_int_value};
-        default: throw std::exception();
+    // For a 64-bit operation, we can do direct comparisons.
+    if (is64)
+        return {left_svalue >= *right_interval.lb().number(), left_svalue <= *right_interval.ub().number()};
+
+    // Handle 32-bit comparisons.
+    if (auto rn = right_interval.singleton()) {
+        if (auto size = inv.eval_interval(left_svalue).finite_size()) {
+            // Find the lowest 64-bit svalue whose low 32 bits match the singleton.
+            int64_t lb = inv.eval_interval(left_svalue).lb().number()->cast_to_sint64();
+            int64_t lb_match = ((lb & 0xFFFFFFFF00000000) | rn->cast_to_sint64());
+            if (lb_match < lb) {
+                lb += 0x100000000;
+            }
+
+            // Find the highest 64-bit svalue whose low 32 bits match the singleton.
+            int64_t ub = inv.eval_interval(left_svalue).ub().number()->cast_to_sint64();
+            int64_t ub_match = ((ub & 0xFFFFFFFF00000000) | rn->cast_to_sint64());
+            if (ub_match > ub) {
+                lb -= 0x100000000;
+            }
```

Comment:
This is a bit clever for me. I am not suggesting any change, since the intention is explained clearly, but if you have any idea / reference that may help the reader understand the logic it would be nice.

Reply:
Added additional comments.